clash-cores/src/Clash/Cores/Xilinx/Synchronizer.hs

```diff
@@ -0,0 +1,74 @@
+{-|
+  Copyright   :  (C) 2022 Google Inc
+  License     :  BSD2 (see the file LICENSE)
+  Maintainer  :  QBayLogic B.V. <devops@qbaylogic.com>
+
+  Xilinx specific synchronization primitives.
+-}
+
+{-# LANGUAGE QuasiQuotes #-}
+
+module Clash.Cores.Xilinx.Synchronizer where
+
+import Clash.Annotations.Primitive
+import Clash.Explicit.Prelude
+
+import Data.String.Interpolate (__i)
+
+-- | See https://docs.xilinx.com/r/en-US/ug901-vivado-synthesis/ASYNC_REG
+dualFlipFlopSynchronizer ::
+  forall dom1 dom2 a.
+  ( KnownDomain dom1
+  , KnownDomain dom2
+  , NFDataX a
+  , BitSize a ~ 1 ) =>
+  Clock dom1 ->
+  Clock dom2 ->
+  Signal dom1 a ->
+  Signal dom2 a
+dualFlipFlopSynchronizer clk1 clk2 =
+    dflipflop clk2
+  . dflipflop clk2
+  . unsafeSynchronizer clk1 clk2
+{-# NOINLINE dualFlipFlopSynchronizer #-}
+{-# ANN dualFlipFlopSynchronizer hasBlackBox #-}
+{-# ANN dualFlipFlopSynchronizer (InlineYamlPrimitive [Verilog, SystemVerilog] [__i|
+ BlackBox:
+   kind: Declaration
+   name: Clash.Cores.Xilinx.Synchronizer.dualFlipFlopSynchronizer
+   template: |-
+     // begin dualFlipFlopSynchronizer
+     (* ASYNC_REG = "TRUE" *) reg ~GENSYM[sync_a][0], ~GENSYM[sync_b][1];
```

Comment:
I wonder if we couldn't just alter the `clash-prelude` `dualFlipFlopSynchronizer` to have a black box that says:
```
~IF~VIVADO~THEN(* ASYNC_REG = "TRUE" *) ~FIreg ~GENSYM[sync_a][0], ~GENSYM[sync_b][1];
```

Then you'd just have to specify `-fclash-hdlsyn vivado` which might already be needed for some workaround anyway.

Reply:
In fact it's would be really a maintainer nightmare to go vendor specific here. We would need to move the `asyncFIFOSynchronizer` to Xilinx as well using the Xilinx specific `dualFlipFlopSynchronizer`. If possible I would greatly prefer to add this to the original `asyncFIFOSynchronizer`

Reply:
I think `asyncFifoSynchronizer` might be better off not using another prim for its FF synchroniser anyway. Since it needs to specify additional constraints, it might as well use a private FF chain with those constraints, once we decide how we add those constraints in the first place.

Reply:
We could just dump it in there anyway. Vendors treat unrecognized attributes as comments anyway AFAIK.